internal/client/git.go

```diff
@@ -0,0 +1,192 @@
+package client
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/caarlos0/log"
+	"github.com/goreleaser/goreleaser/internal/artifact"
+	"github.com/goreleaser/goreleaser/internal/commitauthor"
+	"github.com/goreleaser/goreleaser/internal/git"
+	"github.com/goreleaser/goreleaser/internal/pipe"
+	"github.com/goreleaser/goreleaser/internal/tmpl"
+	"github.com/goreleaser/goreleaser/pkg/config"
+	"github.com/goreleaser/goreleaser/pkg/context"
+	"golang.org/x/crypto/ssh"
+)
+
+const defaultSSHCommand = "ssh -i {{ .KeyPath }} -o StrictHostKeyChecking=accept-new -F /dev/null"
+
+type gitClient struct{}
+
+func NewGitUploadClient(ctx *context.Context) (Client, error) {
+	return &gitClient{}, nil
+}
+
+func (gc *gitClient) CreateFile(ctx *context.Context, commitAuthor config.CommitAuthor, repo Repo, content []byte, path, message string) (err error) {
+	key, err := tmpl.New(ctx).Apply(repo.PrivateKey)
+	if err != nil {
+		return err
+	}
+
+	key, err = keyPath(key)
+	if err != nil {
+		return err
+	}
+
+	var gitSSHCommand string
+	if len(repo.GitSSHCommand) > 0 {
+		gitSSHCommand = repo.GitSSHCommand
+	} else {
+		gitSSHCommand = defaultSSHCommand
+	}
+
+	sshCmd, err := tmpl.New(ctx).WithExtraFields(tmpl.Fields{
+		"KeyPath": key,
+	}).Apply(gitSSHCommand)
+	if err != nil {
+		return err
+	}
+
+	url, err := tmpl.New(ctx).Apply(repo.GitURL)
+	if err != nil {
+		return err
+	}
+
+	if url == "" {
+		return fmt.Errorf("git_url is empty")
+	}
+
+	msg, err := tmpl.New(ctx).Apply(message)
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("%v\n", commitAuthor)
+
+	author, err := commitauthor.Get(ctx, commitAuthor)
+	if err != nil {
+		return err
+	}
+
+	env := []string{fmt.Sprintf("GIT_SSH_COMMAND=%s", sshCmd)}
+
+	repoName := strings.TrimSuffix(url[strings.LastIndex(url, "/")+1:], ".git")
+
+	_, err = os.Stat(filepath.Join(ctx.Config.Dist, repoName))
+
+	// Only clone if this repo has not been cloned before
+	if os.IsNotExist(err) {
```

Comment:
why do we need this?

Reply:
You cannot git clone into a non-empty folder